### DISCUSSION TITLE
Mingw compilation

### DISCUSSION THREAD

**Discussion by pawelsalawa** [utelle/SQLite3MultipleCiphers#24] [General]
Hi,
I'm SQLiteStudio developer. I'd like to express my appreciation for a great work done with this project!

Another thing I wanted to mention is when I tried to compile the amalgamation under Windows using Mingw (64 bit) I received error `error: inlining failed in call to always_inline '_mm_aesimc_si128': target specific option mismatch _mm_aesimc_si128 (__m128i __X)`. After a while I figured out the solution - I need to add `-maes` flag to the compiler.

I know the project is prepared for compilation with VS code, but perhaps this could be useful to you or anyone else seeking for the answer on this error, so I'm just leaving it here, so the Google can index it for them :)

Cheers!

**Comment by utelle** [comment]
Thanks for your feedback.

> [...] I know the project is prepared for compilation with VS code [...]

It is true that the project is mainly developed under Windows using VC++. However, the project provides also build files for MinGW under Windows and for the autoconf toolchain under Linux and macOS.  The latter is tested via CI on each push to the GitHub repository. The build files coming with the project use the GCC compiler option **-march=native** on Windows and **-msse4.2 -maes** on Linux/macOS. Option **-march=native** worked for me when compiling with MinGW, but using options **-msse4.2 -maes** should work just as well or may even be preferrable.

Probably I should add a short note to the amalgamation archive that these compiler flags are required for GCC resp MinGW.

**Reply by little-brother** [reply]
> -march=native

With this option the result can be running only on the same processor.  This is unwanted behaviour in the most cases. `-msse4 -maes` is better as default options.

**Reply by utelle** [reply]
@little-brother, thanks for your comment. I will adjust the MinGW build files for the next release.

**Reply by little-brother** [reply]
mingw32 (under Win7x64) can't build dll :(

> cmd > gcc -shared sqlite3mc.c -o sqlite3.dll -DSQLITE_ENABLE_DBSTAT_VTAB -DSQLITE_ENABLE_COLUMN_METADATA -DSQLITE_THREADSAFE=2 -DSQLITE_ENABLE_FTS4 -DSQLITE_ENABLE_FTS5 -DSQLITE_ENABLE_GEOPOLY -DSQLITE_ENABLE_RTREE -D_WINDOWS -DWIN32 -DNDEBUG -DSQLITE_API=__declspec(dllex
port) -s -O2 -static -msse4 -maes

> c:/mingw/bin/../lib/gcc/mingw32/9.2.0/../../../../mingw32/bin/ld.exe: C:\Temp\ccsfW1p6.o:sqlite3mc.c:(.text+0x57a41): undefined reference to `SystemFunction036@8'
c:/mingw/bin/../lib/gcc/mingw32/9.2.0/../../../../mingw32/bin/ld.exe: C:\Temp\ccsfW1p6.o:sqlite3mc.c:(.text+0x57a60): undefined reference to `SystemFunction036@8'
> collect2.exe: error: ld returned 1 exit status

P.S. mingw64-32 works fine.

**Reply by utelle** [reply]
I suppose you are using the compiler _TDM-GCC 9.2.0_. This is unfortunately a shortcoming of this compiler distribution.

`SystemFunction036` is the entry name of the function `RtlGenRandom`. The function implementation resides in the system DLL `advapi32`, but the link library `advapi32.lib` coming with _TDM-GCC_ does not have the entry point for it. You have to ask the persons responsible for the _TDM-GCC_ distribution, why they decided to not export all functions of the `advapi32` DLL.

I'm aware of this problem and I intend to provide a workaround for it in the future, but right now the only solution is to use a different compiler.

**Reply by utelle** [reply]
@little-brother: 
Commit 07ceacee1a53772203e4302f8cc6f3b4095a73fb changes the MinGW compile option from **-march=native** to **-msse4.2 -maes**. 
A fix for the `RtlGenRandom` (aka `SystemFunction036`) problem is included in [release 1.3.0](https://github.com/utelle/SQLite3MultipleCiphers/releases/tag/v1.3.0).
